### ISSUE TITLE
Injecting a list of dependencies

### ISSUE THREAD

**Issue by JarnoRFB** [ets-labs/python-dependency-injector#243]
Hi, thanks for the nice and well documented framework! There is one point I am currently missing. Let's say I have a class that depends on a list of other classes, I want to achieve something like the following example (written in Python 3.7)

```python
from typing import *
from dependency_injector import containers, providers


class B:
    def __init__(self, val: int):
        self.val = val

    def __repr__(self):
        return f"B({self.val})"


class A:
    def __init__(self, bs: List[B]):
        self.bs = bs

    def __repr__(self):
        return f"A({self.bs})"


def main() -> None:
    class IocContainer(containers.DeclarativeContainer):
        bs = providers.Provider(
            list,
            providers.Factory(B, 1),
            providers.Factory(B, 2),

        )

        a = providers.Factory(
            A,
            bs,
        )

    container = IocContainer()

    print(container.a())


if __name__ == "__main__":
    main()
```

However, using a list of providers seems not to be possible, as the list constructor only takes a single argument, but directly instantiating a list of providers will not instantiate the providers. 
What would be the way to go to inject a list of dependencies.

**Comment by yandrieiev** [comment]
Funnily enough I was exploring this very issue at this very moment. So, I don't know how authors intended to handle such case but it is very easy to circumvent apparent framework limitation - just use `to_list` function instead of `list` as the first argument to your factory.

```
def to_list(*vargs):
    return vargs
```

**Comment by JarnoRFB** [comment]
@yandrieiev thanks for the hint. One could also generalize this for other types that take iterables in their contructor.

```python
from typing import *

from dependency_injector import containers, providers


class B:
    def __init__(self, val: int):
        self.val = val

    def __repr__(self):
        return f"B({self.val})"


class A:
    def __init__(self, bs: List[B]):
        self.bs = bs

    def __repr__(self):
        return f"A({self.bs})"


def collection_of(collection_type, *args):
    return collection_type(args)


def main() -> None:
    class IocContainer(containers.DeclarativeContainer):
        bs = providers.Factory(collection_of, list, providers.Factory(B, 1), providers.Factory(B, 2))

        a = providers.Factory(A, bs)

    container = IocContainer()

    print(container.a())


if __name__ == "__main__":
    main()
```

Probably would be a good idea to document this.

**Comment by rmk135** [comment]
Hello gentlemen,

I would adapt particular case to:

```python
from typing import *

from dependency_injector import containers, providers


class B:
    def __init__(self, val: int):
        self.val = val

    def __repr__(self):
        return f"B({self.val})"


class A:
    def __init__(self, *bs: List[B]):
        self.bs = bs

    def __repr__(self):
        return f"A({self.bs})"


def main() -> None:
    class IocContainer(containers.DeclarativeContainer):

        a = providers.Factory(
            A,
            providers.Factory(B, 1),
            providers.Factory(B, 2),
        )

    container = IocContainer()

    print(container.a())


if __name__ == "__main__":
    main()
```

The solution is not identical to provided ones. There are two differences:

1. It limits `A` from having other positional arguments.
2. Context positional arguments will be appended (details at bullet #4 at http://python-dependency-injector.ets-labs.org/providers/factory.html#factory-providers-and-init-injections).

I like the workaround proposed by @yandrieiev and `collection_of(...)` solution looks accurate @JarnoRFB . This feature is a good release candidate. I will consider building it in.

Roman

**Comment by rmk135** [comment]
@JarnoRFB @yandrieiev 

I have created a new provider - `List`. It can be used for injecting a list of values. It's as simple as this:

```python
dispatcher_factory = providers.Factory(
    Dispatcher,
    modules=providers.List(
        providers.Factory(Module, name='m1'),
        providers.Factory(Module, name='m2'),
    ),
)
```

The original example from this issue now can look like this:


```python
class IocContainer(containers.DeclarativeContainer):
    bs = providers.List(
        providers.Factory(B, 1),
        providers.Factory(B, 2),

    )

    a = providers.Factory(
        A,
        bs,
    )
```

`List` provider is available in version `3.16.0`. Version `3.16.0` is published on [PyPI](https://pypi.org/project/dependency-injector/). You're welcome to upgrade.

Roman
